### PR TITLE
Refactoring ValidatorBase

### PR DESCRIPTION
**Pull request by eddynaka**

### CODE REVIEW

**Comment by michaelcfanning** [comment]
>         public static bool TryGetNonEmptyValue<TKey>(this Dictionary<TKey, FlexMatch> dictionary, TKey key, out FlexMatch value)

Are you sure you require these helpers now? Trying removing them. i.e., any Dictionary instance should resolve to the IDictionary helpers below, I believe.

---
Refers to: Src/Plugins/Security/Utilities/DictionaryExtensions.cs:12 in 3b14c3c. [](commit_id = 3b14c3c71bd80e6067ec75be3793f0fef620ea11, deletion_comment = False)

File: Src/Plugins/Tests.Security/Validators/SEC101_001.HttpAuthorizationRequestHeaderValidatorTests.cs [hunk 1]
```diff
@@ -141,6 +140,7 @@ public void HttpAuthorizationRequestHeaderValidator_MockHttpTests()
 
             var sb = new StringBuilder();
 
+            var httpAuhorizationRequestHeaderValidator = new HttpAuthorizationRequestHeaderValidator();
```
Comment (michaelcfanning) [inline Src/Plugins/Tests.Security/Validators/SEC101_001.HttpAuthorizationRequestHeaderValidatorTests.cs:143]: >httpAuhorizationRequestHeaderValidator [](http://example.com/codeflow?start=16&length=38)

This is misspelled now.

S/be httpAuthorizationRequestionHeaderValidator

File: Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/ExpectedOutputs/SEC102_001.EmailAddress_with_email_addresses.sarif [hunk 2]
```diff
@@ -283,47 +283,6 @@
...
-              }
-            }
-          ],
-          "fingerprints": {
-            "AssetFingerprint/v1": "[host=example.com][id=lastname]",
-            "ValidationFingerprint/v1": "[host=example.com][id=lastname]",
-            "ValidationFingerprintHash/v1": "a82a08b8f8cd5a7eabf30386974158e85fa1f0982b13d273aa94373c2638eb42",
-            "AssetFingerprint/v2": "{\"host\":\"example.com\",\"id\":\"lastname\"}",
-            "ValidationFingerprint/v2": "{\"host\":\"example.com\",\"id\":\"lastname\"}"
-          }
```
Comment (michaelcfanning) [inline Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/ExpectedOutputs/SEC102_001.EmailAddress_with_email_addresses.sarif:325]: >} [](http://example.com/codeflow?start=10&length=1)

Why has this stuff gone away?

File: Src/Test.UnitTests.Sarif.PatternMatcher.Function/HttpAnalyzeFunctionTests.cs [hunk 3]
```diff
@@ -51,7 +51,7 @@ public async Task Function_HttpAnalyze_FileWithPAT_Should_Return_SarifLog()
                 log: logger,
                 context: TestHelper.ContextSetup());
 
-            ValidateResult(string.Empty, result, runCount: 1, resultCount: 2, FailureLevel.Note, ignoreRegionContent: true);
+            ValidateResult(string.Empty, result, runCount: 1, resultCount: 1, FailureLevel.Note, ignoreRegionContent: true);
```
Comment (michaelcfanning) [inline Src/Test.UnitTests.Sarif.PatternMatcher.Function/HttpAnalyzeFunctionTests.cs:54]: >ValidateResult [](http://example.com/codeflow?start=12&length=14)

Why did this test result change?

File: Src/Sarif.PatternMatcher.Sdk/StaticValidatorBase.cs [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+using System.Diagnostics;
+using System.Net;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+
+using Microsoft.RE2.Managed;
+
+namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher.Sdk
+{
+    public abstract class StaticValidatorBase
```
Comment (michaelcfanning) [inline Src/Sarif.PatternMatcher.Sdk/StaticValidatorBase.cs:16]: >StaticValidatorBase [](http://example.com/codeflow?start=26&length=19)

Nice rename! Good choice.

File: Src/Plugins/Tests.Security/Validators/SEC101_006.GitHubPatValidatorTests.cs [hunk 5]
```diff
@@ -24,7 +24,9 @@ public void GitHubPatValidator_TestStatic()
             groups.Add("checksum", new FlexMatch() { Value = "checksum" });
             groups.Add("scanTargetFullPath", new FlexMatch() { Value = "GitHitPatTest" });
 
-            IEnumerable<ValidationResult> validationResults = GitHubPatValidator.IsValidStatic(groups);
+            var gitHubPatValidator = new GitHubPatValidator();
```
Comment (michaelcfanning) [inline Src/Plugins/Tests.Security/Validators/SEC101_006.GitHubPatValidatorTests.cs:27]: >var [](http://example.com/codeflow?start=12&length=3)

in other rules, you don't introduce a vertical line after instantiating the rule. :) Style nit!

File: Src/Plugins/Tests.Security/TestData/SecurePlaintextSecrets/ExpectedOutputs/SEC101_102.AdoPat.sarif [hunk 6]
```diff
@@ -81,51 +81,6 @@
...
-              "",
-              "Azure DevOps personal access token",
-              "",
-              ""
-            ]
-          },
-          "locations": [
-            {
-              "physicalLocation": {
-                "artifactLocation": {
```
Comment (michaelcfanning) [inline Src/Plugins/Tests.Security/TestData/SecurePlaintextSecrets/ExpectedOutputs/SEC101_102.AdoPat.sarif:103]: >t [](http://example.com/codeflow?start=29&length=1)

why did this go away? In general, it's good to explicitly account for results that disappear. 

We have either introduced new false negatives or you have potentially dropped false positives, right?

Where's the description, if so, in ReleaseHistory.md?

**Review by michaelcfanning — APPROVED** [review]
:shipit:

File: Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/ExpectedOutputs/SEC102_001.EmailAddress_with_email_addresses.sarif (hunk 2 above)
Comment (eddynaka) [reply Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/ExpectedOutputs/SEC102_001.EmailAddress_with_email_addresses.sarif:325]: we weren't using ValidatorBase, so duplicated results would appear.
With my current change, we will ask for the static or dynamic validator class, removing duplications

File: Src/Test.UnitTests.Sarif.PatternMatcher.Function/HttpAnalyzeFunctionTests.cs (hunk 3 above)
Comment (eddynaka) [reply Src/Test.UnitTests.Sarif.PatternMatcher.Function/HttpAnalyzeFunctionTests.cs:54]: Before this change, AdoPat validator didn't implemented the ValidatorBase generating duplicated results.
With my current change, we will remove duplicates

File: Src/Plugins/Tests.Security/TestData/SecurePlaintextSecrets/ExpectedOutputs/SEC101_102.AdoPat.sarif (hunk 6 above)
Comment (eddynaka) [reply Src/Plugins/Tests.Security/TestData/SecurePlaintextSecrets/ExpectedOutputs/SEC101_102.AdoPat.sarif:103]: would appear.
With my current change, we will ask for the static or dynamic validator class, removing duplications

**Pull request merged**